pytensor/graph/basic.py

```diff
@@ -970,6 +970,134 @@ def applys_between(
+
+    Parameters
+    ----------
+    outputs : Collection[Variable]
+        Variable to get conditions for
+    conditions : Optional[Collection[Variable]]
+        Additional conditions to assume, by default None
+
+    Returns
+    -------
+    List[Variable]
+        Variables required to compute ``outputs``
+
+    Examples
+    --------
+    The returned nodes marked in (parenthesis), condition nodes are ``c``, output nodes are ``o``
+
+    * No conditions
+
+    .. code-block::
+
+        n - n - (o)
+
+    * One condition
+
+    .. code-block::
+
+        n - (c) - o
+
+    * Two conditions where on depends on another, both returned
+
+    .. code-block::
+
+        (c) - (c) - o
+
+    * Additional nodes are present
+
+    .. code-block::
+
+           (c) - n - o
+        n - (n) -'
+
+    * Disconnected condition not returned
+
+    .. code-block::
+
+        (c) - n - o
+         c
+
+    * Disconnected output is present and returned
+
+    .. code-block::
+
+        (c) - (c) - o
+        (o)
+
+    * Condition on itself adds itself
+
+    .. code-block::
+
+        n - (c) - (o/c)
+
+    """
+    # simple case, no additional conditions
+    blockers: Set[Variable] = set()
+    independent_nodes = list()
+    # blockers have known independent nodes and pre-conditions
+    candidates = list(outputs)
+    if not conditions:  # None or empty
+        # just filter out unique variables
+        for node in candidates:
+            if node not in blockers:
+                independent_nodes.append(node)
+                blockers.add(node)
+        # no more actions are needed
+        return independent_nodes
+    blockers.update(conditions)
+    # enforce O(1) check for node in conditions
+    conditions = blockers.copy()
+    # track conditions that are inside the graph to remove disconnected later
+    # preserve order for return
+    conditions_inside = list()
+
+    while candidates:
+        # on any new candidate
+        node = candidates.pop()
+        # check if the node is independent, never go above blockers
+        # blockers are independent nodes and pre-conditions
+        if node in conditions:
+            # The case where node is in conditions so we check if it depends on others
+            # it should be removed from the blockers to check against the rest
+            dependent = variable_is_in_ancestors(node, blockers - {node})
+            # conditions that are present in the graph (not disconnected)
+            # should be added to conditions_inside
+            conditions_inside.append(node)
+            if dependent:
+                # if the condition is still dependent we need to go above,
+                # the search is not yet finished
+                # the node _has_ to have owner to be dependent
+                # so we do not check it
+                # and populate search to go above
+                candidates.extend(node.owner.inputs)
```

Comment:
Could this inputs have already been checked/ruled out?

Reply:
I can add a check to filter out those who are in blockers though, would be also correct

Reply:
There's also the possibility of repeated inputs, so you may want to extend in a loop. But if this does not cause problems as is, we can have a bit more redundancy for the sake of simpler code and revisit later if it turns out to be a bottleneck

Reply:
Just checked, it breaks the test for replacements, so I decided to leave it as is